### PR TITLE
Add Top-nσ logit truncation custom logit processor example

### LINKED ISSUE #29397: Add Top-nσ logit truncation custom logit processor example
## Motivation

SGLang's custom logit processor framework (enabled via `--enable-custom-logit-processor`) lets users inject arbitrary logit-space processing per-request. However, the only existing reference is a **dummy** `DeterministicLogitProcessor` in the docs that always samples a fixed token — there is no example of a processor that does meaningful logit filtering.

This issue proposes adding the **first real-world custom logit processor example**: [Top-nσ](https://arxiv.org/abs/2411.07641) ("Top-nσ: Not All Logits Are You Need", ACL 2025), a logit-space dynamic truncation method.

```
threshold = max_logit - n * std_logit
logits[logits < threshold] = -inf
```

Unlike sglang's built-in probability-space filters (`top_k`, `top_p`, `min_p`), Top-nσ operates **before softmax** (where SGLang's custom processor pipeline runs) and adapts the candidate set to the "peakiness" of the distribution.

### How it compares to existing methods

| Method | Space | Truncation rule | Adaptive? |
|---|---|---|---|
| top_k | Probability | Keep top-k highest-probability tokens | ❌ Fixed count |
| top_p (nucleus) | Probability | Keep tokens until cumulative prob ≥ p | ✅ Adaptive count only |
| min_p | Probability | Keep tokens with prob ≥ max_prob × min_p | ✅ Relative threshold |
| top_nσ | **Logit** | threshold = max_logit − n × std | ✅ Dual-adaptive |

> Key distinction: sglang's top_k/top_p/min_p all operate **after softmax** (probability space); top_nσ operates **before softmax** (logit space). Top-nσ is the first logit-space truncation reference for the custom processor framework.

### Why Top-nσ?

- **Truncation before softmax** — filters in logit space first, then softmax only processes the surviving candidates.
- **Dual adaptivity** — the std-based threshold adjusts both *how many* and *how loosely* candidates are kept in one formula.
- **Temperature-stable** — candidate count stays consistent across temperature scaling (logit std scales proportionally).
- **Argmax-invariant** — the top-1 token always survives, so greedy decoding is unaffected.

## Proposal

Add an example in a new `examples/runtime/custom_logit_processor/` folder — `top_n_sigma.py` — following the existing `examples/runtime/` pattern (standalone script, documented server launch command, client-side usage).

**No framework changes**: the processor is defined in the example file and injected per-request via the existing `custom_logit_processor` field + `--enable-custom-logit-processor` gate.

### Implementation characteristics

- **Fully vectorized** across the batch (no per-row Python loop)
- **std computed once** and reused for both the `std==0` guard and the threshold formula (avoids a duplicate vocab-wide reduction)
- **Argmax-invariant** by construction

### Edge cases handled

| Case | Behavior |
|------|----------|
| `top_n_sigma <= 0` | Request skipped |
| `top_n_sigma` is None / non-numeric | Request skipped |
| `std == 0` (all logits equal) | Row skipped (no filtering) |
| NaN/Inf logits | Row skipped (finite guard) |
| Empty / None param list | Logits returned unchanged |

## References

Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### PR DESCRIPTION
**Pull request by Jackie2049**
## Summary

Closes #29397

Adds the **first real-world custom logit processor example** to `examples/runtime/custom_logit_processor/`, demonstrating [Top-nσ](https://arxiv.org/abs/2411.07641) (Tang et al., ACL 2025) logit truncation.

The existing docs only show a dummy `DeterministicLogitProcessor`; this example gives users a complete, working reference for a custom processor that does meaningful logit-space filtering.

Based on: Tang et al., "Top-nσ: Not All Logits Are You Need", ACL 2025. [arXiv:2411.07641](https://arxiv.org/abs/2411.07641)

### What is Top-nσ?

A logit-space dynamic truncation method that uses the standard deviation of the logit distribution to set a filtering threshold:

```
threshold = max_logit - n * std_logit
logits[logits < threshold] = -inf
```

It operates **before softmax** (exactly where SGLang's custom logit processor pipeline runs — `_preprocess_logits` is called before temperature scaling) and adapts to distribution "peakiness":
- **Sharp distribution** (model certain): small std → narrow threshold → fewer candidates
- **Flat distribution** (model uncertain): large std → wide threshold → more candidates

### How it compares to existing truncation methods

| Method | Space | Truncation rule | Adaptive? |
|---|---|---|---|
| top_k | Probability | Keep top-k highest-probability tokens | ❌ Fixed count |
| top_p (nucleus) | Probability | Keep tokens until cumulative prob ≥ p | ✅ Adaptive count only |
| min_p | Probability | Keep tokens with prob ≥ max_prob × min_p | ✅ Relative threshold |
| top_nσ | **Logit** | threshold = max_logit − n × std | ✅ Dual-adaptive |

> Key distinction: sglang's top_k/top_p/min_p all operate **after softmax** (probability space); top_nσ operates **before softmax** (logit space).

## Changes

- **New example**: `examples/runtime/custom_logit_processor/top_n_sigma.py` — the processor definition + `/generate` and OpenAI chat completion usage

**No changes to framework code** — `SamplingParams`, the sampler, `custom_logit_processor.py`, and the README are all untouched. The processor is defined in the example file and injected per-request via the existing `custom_logit_processor` field.

### Processor implementation highlights

- **Fully vectorized** across the batch (no Python per-row loop)
- **std computed once** and reused for both the `std==0` guard and the threshold formula, avoiding a duplicate vocab-wide reduction
- **Argmax-invariant** by construction (the peak token always survives)

### Edge cases handled

| Case | Behavior |
|------|----------|
| `top_n_sigma <= 0` | Request skipped |
| `top_n_sigma` is None / non-numeric | Request skipped |
| `std == 0` (all logits equal) | Row skipped (no filtering) |
| NaN/Inf logits | Row skipped (finite guard) |
| Empty / None param list | Logits returned unchanged |

## Usage

```bash
# 1. Launch the server with the custom-logit-processor gate enabled
python -m sglang.launch_server \
    --model meta-llama/Meta-Llama-3-8B-Instruct \
    --enable-custom-logit-processor

# 2. Run the example
python examples/runtime/custom_logit_processor/top_n_sigma.py
```

The processor is serialized via `CustomLogitProcessor.to_str()` and passed per-request:

```python
import requests
from sglang.srt.sampling.custom_logit_processor import CustomLogitProcessor

class TopNSigmaLogitProcessor(CustomLogitProcessor):
    # ... (see example file for the full implementation)
    pass

response = requests.post("http://localhost:30000/generate", json={
    "text": "The capital of France is",
    "custom_logit_processor": TopNSigmaLogitProcessor.to_str(),
    "sampling_params": {
        "temperature": 0.8,
        "custom_params": {"top_n_sigma": 2.0},
    },
})
```

## Testing plan

- [x] Example file parses without syntax errors
- [x] Processor logic verified (vectorization, argmax invariance, edge cases)
- [ ] Manual end-to-end test with `--enable-custom-logit-processor` on a small model
- [ ] CI passes

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #28990032152](https://github.com/sgl-project/sglang/actions/runs/28990032152)
Latest PR Test (Extra): :x: [Run #28990032069](https://github.com/sgl-project/sglang/actions/runs/28990032069)

### CODE REVIEW

**Comment by Codekiing** [comment]
Could someone with triage access add the `run-ci` label to trigger CI? The code fixes (GPU copy optimization + `openai.OpenAI`) have been merged into this branch via Jackie2049/sglang#4 and the bot review comments are now outdated. Just need the label to run CI. Thanks!

File: examples/runtime/custom_logit_processor/top_n_sigma.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        max_logits = selected.max(dim=-1, keepdim=True).values
+        thresholds = max_logits - n_sigmas_t.unsqueeze(-1) * std_all
+
+        # Mask tokens below threshold, only on rows that pass the guards.
+        selected.masked_fill_(
+            (selected < thresholds) & process_mask, float("-inf")
+        )
+        # Advanced indexing produced a copy; write the modified rows back.
+        logits[rows_t] = selected
+        return logits
```
[the remark this answers is not in the document]
Comment (Codekiing) [reply examples/runtime/custom_logit_processor/top_n_sigma.py]: Fixed in commit 771a156: replaced advanced indexing with in-place boolean mask on the full logits tensor, avoiding GPU copies.

File: examples/runtime/custom_logit_processor/top_n_sigma.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    ).json()
+    print("=== /generate with top_n_sigma=2.0 ===")
+    print(response["text"])
+
+
+def run_openai():
+    """OpenAI-compatible chat completions endpoint."""
+    import openai
+
+    client = openai.Client(base_url=url + "/v1", api_key="None")
```
[the remark this answers is not in the document]
Comment (Codekiing) [reply examples/runtime/custom_logit_processor/top_n_sigma.py]: Fixed in commit 771a156: changed `openai.Client` to `openai.OpenAI`.

**Comment by ajinkyajawale14499** [comment]
Hi @Jackie2049 — thanks, and sorry for the duplicate! Yours was up first, so let's ship #29398. I'm closing my #29473 in favor of this.

A few concrete improvements I can contribute:

1. **fp32 reductions** — compute `std`/threshold in fp32. Right now `logits.std(dim=-1)` runs in `logits.dtype`, which can be unstable for fp16/bf16 logits; casting to fp32 for the reduction (then masking the original) is safer.
2. **Reject `bool` in validation** — `isinstance(True, int)` is `True`, so `top_n_sigma=True` currently slips through as `n=1.0`. A small `isinstance(n, bool)` guard fixes it.
3. **Demo + test** — optionally a baseline-vs-top-nσ comparison in the `/generate` example, and an end-to-end smoke test against a small model with `--enable-custom-logit-processor`.

I verified the core algorithm locally (argmax-invariance, threshold math, std==0 / NaN-Inf / invalid-param / empty-list / per-row mixed-n) so I can fold those edge cases in too.

How would you like them — a PR into your branch, review suggestions, or direct commits (happy to be added as a collaborator on your fork)? Either way, glad to co-author. 🙌

**Comment by ajinkyajawale14499** [comment]
@Jackie2049 opened a small PR into your branch with the two correctness fixes (fp32 std/max reductions + bool guard), verified locally (8/8, incl. fp16 stability): Jackie2049/sglang#5 — merge or cherry-pick at will. Happy to add the end-to-end test next if useful.

**Comment by ajinkyajawale14499** [comment]
Added the end-to-end test too (in Jackie2049/sglang#5): `TestSRTEndpoint.test_top_n_sigma_argmax_invariant` 

- greedy (temperature=0) generation with vs without the processor must produce identical tokens, since Top-nσ never removes the max-logit token. 
- It reuses the existing `--enable-custom-logit-processor` server in that test class (1-gpu-small CUDA CI), so no new server launch. 
- ruff-clean. (Couldn't run the GPU test locally — Apple Silicon — so CI is the gate there.)

**Comment by Jackie2049** [comment]
@ajinkyajawale14499 Thanks for the heads-up and for jumping in — really appreciate you contributing to this! I'll review Jackie2049/sglang#5 right away (the fp32 reductions, bool guard, and the end-to-end test).

**Comment by Jackie2049** [comment]
@ajinkyajawale14499 I genuinely appreciate your interest in this proposal — it's been a pleasure seeing you engage with the details and put together meaningful improvements.

I've carefully reviewed your PR (#5) and merged it into our branch. Your contributions are all well thought out and have meaningfully improved this example.

Looking ahead, I'd like to invite you to stay involved as we continue to refine this PR and work toward getting it merged upstream. There's still ground to cover — further polish, community review, and building broader attention. I'd be happy to have you as a co-contributor as we push this forward.

Let's keep at it — I'm looking forward to shipping this together.
